File: assets/explainer.py
"""
Explainer — zero-deps Python stdlib — Kernel SHAP + LIME tabular + narrative
LCG 20260813→189831298 idx3820 triple[11205,19448,14209] same-link-same-stars
No pip shap/lime — pure stdlib math + random
Usage: from explainer import explain_prediction
"""
import math, random, datetime, json

RAND_A=1103515245
RAND_C=12345
_R=20260813 ^ 189831298

def _rnd():
    global _R
    _R = (_R*1103515245+12345) & 0x7fffffff
    return (_R+1)/2147483648.0

def randn():
    # Box-Muller
    u = 0.0
    while u==0:
        u=_rnd()
    v=0.0
    while v==0:
        v=_rnd()
    return math.sqrt(-2*math.log(u))*math.cos(2*math.pi*v)

def comb(n,k):
    if k<0 or k>n:
        return 0
    if k>n-k:
        k=n-k
    c=1.0
    for i in range(1,k+1):
        c=c*(n-k+i)/i
        if not math.isfinite(c):
            return 1e12
    return c

def shap_kernel(s,M):
    if s==0 or s==M:
        return 1000.0
    cv=comb(M,s)
    if cv==0:
        return 1.0
    return (M-1)/(cv * s * (M-s))

def weighted_leastsq(X,y,w,ridge=1e-6):
    N=len(X)
    D=len(X[0])
    A=[[0.0]*D for _ in range(D)]
    b=[0.0]*D
    for i in range(N):
        wi=w[i]
        xi=X[i]
        yi=y[i]
        for a in range(D):
            b[a]+=wi*xi[a]*yi
            for bb in range(D):
                A[a][bb]+=wi*xi[a]*xi[bb]
    if ridge:
        for d in range(D):
            A[d][d]+=ridge
    # augment
    Mmat=[row[:] + [bv] for row,bv in zip(A,b)]
    for col in range(D):
        prow=col
        for r in range(col,D):
            if abs(Mmat[r][col])>abs(Mmat[prow][col]):
                prow=r
        if abs(Mmat[prow][col])<1e-12:
            continue
        if prow!=col:
            Mmat[prow],Mmat[col]=Mmat[col],Mmat[prow]
        piv=Mmat[col][col]
        for j in range(col,D+1):
            Mmat[col][j]/=piv
        for r in range(D):
            if r==col:
                continue
            fac=Mmat[r][col]
            if abs(fac)<1e-12:
                continue
            for j in range(col,D+1):
                Mmat[r][j]-=fac*Mmat[col][j]
    return [r[D] for r in Mmat]

def explain_prediction(x, feature_names, predict_fn=None, baseline=None, opts=None):
    opts=opts or {}
    M=len(x)
    names=feature_names if feature_names and len(feature_names)==M else [f"f{i}" for i in range(M)]
    base = baseline if baseline and len(baseline)==M else [0.0]*M
    num_shap=min(opts.get("numShap", max(64,2*M+24)),160)
    num_lime=min(opts.get("numLime",96),256)
    kw=opts.get("kernelWidth", math.sqrt(M)*0.75)
    sigma=opts.get("sigma",1.0)
    domain=opts.get("domain","generic")

    def pred_wrapper(vec):
        if predict_fn:
            return predict_fn(vec)
        # linear fallback avg
        return sum(vec)/M

    pred=pred_wrapper(x)
    f_base=pred_wrapper(base)

    Xsh=[]; ysh=[]; wsh=[]
    for k in range(num_shap):
        if k==0:
            mask=[0]*M
        elif k==1:
            mask=[1]*M
        else:
            # random size
            if _rnd()<0.22:
                sz=1 if _rnd()<0.5 else M-1
            else:
                sz=int(1+_rnd()*(M-1))
                if sz<1: sz=1
                if sz>=M: sz=M-1
            mask=[0]*M
            idxs=list(range(M))
            # partial shuffle
            for i in range(sz):
                j=i+int(_rnd()*(M-i))
                idxs[i],idxs[j]=idxs[j],idxs[i]
                mask[idxs[i]]=1
            if sum(mask)!=sz:
                # fallback ensure — mask already sized, but ensure not empty
                pass
        sz=sum(mask)
        w=shap_kernel(sz,M)
        z=[ x[i] if mask[i] else base[i] for i in range(M) ]
        y=pred_wrapper(z)
        row=[1]+mask
        Xsh.append(row); ysh.append(y); wsh.append(w)

    beta_sh=weighted_leastsq(Xsh,ysh,wsh,ridge=1e-6)
    intercept=beta_sh[0]
    shap_arr=beta_sh[1:]
    sum_shap=sum(shap_arr)
    exp_full=intercept+sum_shap
    fidelity=abs(pred-exp_full)/(abs(pred)+1e-6)

    # LIME
    Xli=[]; yli=[]; wli=[]
    for i in range(num_lime):
        z=[ x[j]+randn()*sigma*0.35 for j in range(M) ]
        d2=sum((z[j]-x[j])**2 for j in range(M))
        w=math.sqrt(math.exp(-d2/(kw*kw)))
        y=pred_wrapper(z)
        Xli.append([1]+z); yli.append(y); wli.append(w)

    beta_lime=weighted_leastsq(Xli,yli,wli,ridge=1.0)
    lime_intercept=beta_lime[0]
    lime_coefs=beta_lime[1:]

    shap={names[i]:shap_arr[i] for i in range(M)}
    lime={names[i]:lime_coefs[i] for i in range(M)}

    # narrative
    def topk(vals,k=3):
        iv=sorted([(abs(vals[i]),i,vals[i]) for i in range(M)], reverse=True)[:k]
        return [{"name":names[i],"val":v,"abs":a,"idx":i} for a,i,v in iv]

    tops=topk(shap_arr,3)
    lime_tops=topk(lime_coefs,3)
    def fmt(v): return f"+{v:.2f}" if v>=0 else f"{v:.2f}"
    shap_str=", ".join([f"{t['name']} (SHAP {fmt(t['val'])})" for t in tops])
    delta=pred-f_base
    lime_check = f"LIME locally {lime_tops[0]['name']} {'pushes up' if lime_tops[0]['val']>=0 else 'pulls down'} ({fmt(lime_tops[0]['val'])})" if lime_tops else ""
    pov={
        "owner": f"Ownership: projects {fmt(delta)} vs baseline because {shap_str}.",
        "player": f"Fit: {lime_check} — secondary {tops[1]['name'] if len(tops)>1 else 'stable'} {fmt(tops[1]['val']) if len(tops)>1 else ''} situational.",
        "brand": f"Story: headline driver {tops[0]['name'] if tops else 'unknown'} ({fmt(tops[0]['val']) if tops else '0'}) lifts brand when {lime_tops[0]['name'] if lime_tops else 'context'} aligns.",
        "dfs": f"DFS: exploitable if {tops[0]['name'] if tops else 'top'} {'remains mispriced' if tops and tops[0]['val']>0 else 'faded'} — SHAP sum {fmt(sum_shap)} → pred {pred:.2f} vs base {f_base:.2f}."
    }
    generic=f"This pick projects {fmt(delta)} vs baseline ({f_base:.2f}→{pred:.2f}) because {shap_str}. {lime_check}. {pov['owner']}"

    return {
        "prediction": pred,
        "baseline": {"value": f_base, "vector": base},
        "shap_values": shap,
        "shap_array": shap_arr,
        "lime_values": lime,
        "lime_array": lime_coefs,
        "intercept_shap": intercept,
        "intercept_lime": lime_intercept,
        "fidelity": {"shap_additive_error": abs(pred-exp_full), "shap_relative": fidelity, "sum_shap": sum_shap, "expected": exp_full},
        "stability": {"numShap": num_shap, "numLime": num_lime, "kernelWidth": kw},
        "feature_names": names,
        "narrative": {"owner":pov["owner"],"player":pov["player"],"brand":pov["brand"],"dfs":pov["dfs"],"generic":generic,"tops":tops,"limeTops":lime_tops,"delta":delta},
        "meta": {"M":M,"domain":domain,"timestamp": datetime.datetime.utcnow().isoformat()+"Z","lcg":"20260813→189831298 idx3820 triple[11205,19448,14209]"}
    }

if __name__=="__main__":
    # smoke test
    x=[1,2,3]
    fn=lambda v: v[0]*0.5+v[1]*1.2-v[2]*0.3
    print(json.dumps(explain_prediction(x,["a","b","c"],fn,baseline=[0,0,0],opts={"domain":"hoops"}), indent=2))

File: assets/explainer.js
/**
 * Explainer — zero-deps JS — Kernel SHAP + LIME tabular + narrative
 * SSOT: bundles/coordination/active-tasks.md LCG 20260813→189831298 idx3820 triple[11205,19448,14209] same-link-same-stars
 * Works stdlib-only; no shap/lime pip; model-agnostic.
 * API: explainPrediction(x, featureNames, predictFn, opts) -> {prediction, baseline, shap, lime, additive_fidelity, narrative:{owner,player,brand,dfs,generic}}
 *
 * x: number[] feature vector
 * featureNames: string[]
 * predictFn: (vec:number[])=>number — must be pure sync; if omitted uses linear fallback
 *
 * zero-deps true — no imports
 */
(function (root) {
  const RAND_A = 1103515245, RAND_C = 12345, RAND_M = 0x80000000;
  function lcg(seed){ return (Math.imul(seed, RAND_A) + RAND_C) >>> 0 & 0x7fffffff; }
  // gaussian via box-muller using LCG uniform
  let _s = 20260813 ^ 189831298;
  function rnd(){ _s = lcg(_s); return (_s+1)/0x80000000; }
  function randn(){ let u=0,v=0; while(u===0) u=rnd(); while(v===0) v=rnd(); return Math.sqrt(-2*Math.log(u))*Math.cos(2*Math.PI*v); }

  function comb(n,k){
    if(k<0||k>n) return 0;
    if(k>n-k) k=n-k;
    let c=1;
    for(let i=1;i<=k;i++){ c=c*(n-k+i)/i; if(!isFinite(c)) return 1e12; }
    return c;
  }
  function shapKernel(s,M){
    if(s===0||s===M) return 1000;
    const combv = comb(M,s);
    if(combv===0) return 1;
    return (M-1)/(combv * s * (M-s));
  }

  function weightedLeastSq(X,y,w,ridge){
    // X: N x D, y: N, w: N, ridge lambda
    const N=X.length, D=X[0].length;
    // compute XtWX and XtWy
    const A = Array.from({length:D},()=>Array(D).fill(0));
    const b = Array(D).fill(0);
    for(let i=0;i<N;i++){
      const wi=w[i]; const xi=X[i]; const yi=y[i];
      for(let a=0;a<D;a++){
        b[a]+=wi*xi[a]*yi;
        for(let bb=0;bb<D;bb++){
          A[a][bb]+=wi*xi[a]*xi[bb];
        }
      }
    }
    if(ridge){ for(let d=0;d<D;d++) A[d][d]+=ridge; }
    // solve A x = b via Gauss-Jordan
    // augment
    const M = A.map((row,i)=> row.concat([b[i]]));
    for(let col=0;col<D;col++){
      // pivot
      let prow=col;
      for(let r=col;r<D;r++) if(Math.abs(M[r][col])>Math.abs(M[prow][col])) prow=r;
      if(Math.abs(M[prow][col])<1e-12) continue;
      if(prow!==col){ const tmp=M[prow]; M[prow]=M[col]; M[col]=tmp; }
      const piv=M[col][col];
      for(let j=col;j<=D;j++) M[col][j]/=piv;
      for(let r=0;r<D;r++){
        if(r===col) continue;
        const factor=M[r][col];
        if(Math.abs(factor)<1e-12) continue;
        for(let j=col;j<=D;j++) M[r][j]-=factor*M[col][j];
      }
    }
    const beta = M.map(r=>r[D]);
    return beta;
  }

  function dot(a,b){ let s=0; for(let i=0;i<a.length;i++) s+=a[i]*b[i]; return s;}

  function explainPrediction(x, featureNames, predictFn, opts={}){
    opts = opts||{};
    const M = x.length;
    const names = featureNames && featureNames.length===M ? featureNames : x.map((_,i)=>`f${i}`);
    const baseline = opts.baseline && opts.baseline.length===M ? opts.baseline.slice() : Array(M).fill(0);
    const numShap = Math.min(opts.numShap||Math.max(64, 2*M+24), 160);
    const numLime = Math.min(opts.numLime||96, 256);
    const kernelWidth = opts.kernelWidth || Math.sqrt(M)*0.75;

    const pred = (typeof predictFn==='function') ? predictFn(x) : dot(x, Array(M).fill(1/M));

    // ---- Kernel SHAP ----
    let Xsh=[], ysh=[], wsh=[];
    // add baseline and full
    const fBaseline = (typeof predictFn==='function') ? predictFn(baseline) : 0;
    const fFull = pred;
    // enumerate coalitions
    for(let k=0;k<numShap;k++){
      let mask;
      if(k===0){ mask=Array(M).fill(0); }
      else if(k===1){ mask=Array(M).fill(1); }
      else{
        // random size
        const sz = (rnd()<0.22) ? (rnd()<0.5?1:M-1) : Math.floor(1+ rnd()*(M-1));
        mask=Array(M).fill(0);
        const idxs=[...Array(M).keys()];
        // shuffle partial fisher-yates for sz
        for(let i=0;i<sz;i++){
          const j=i+Math.floor(rnd()*(M-i));
          [idxs[i],idxs[j]]=[idxs[j],idxs[i]];
          mask[idxs[i]]=1;
        }
      }
      const sz = mask.reduce((a,b)=>a+b,0);
      const w = shapKernel(sz,M);
      const z = mask.map((m,i)=> m ? x[i] : baseline[i]);
      const y = (typeof predictFn==='function') ? predictFn(z) : dot(z, Array(M).fill(1/M));
      const row = [1].concat(mask); // intercept + mask
      Xsh.push(row);
      ysh.push(y);
      wsh.push(w);
    }
    const betaSh = weightedLeastSq(Xsh, ysh, wsh, 1e-6);
    const intercept = betaSh[0];
    const shapArr = betaSh.slice(1);
    // additive check
    const sumShap = shapArr.reduce((a,b)=>a+b,0);
    const expFull = intercept + sumShap;
    const fidelity = Math.abs(fFull - expFull) / (Math.abs(fFull)+1e-6);

    // ---- LIME ----
    let Xli=[], yli=[], wli=[];
    // precompute std for scaling if provided or estimate from x spread
    const sigma = opts.sigma || 1.0;
    for(let i=0;i<numLime;i++){
      const z = x.map(v=> v + randn()*sigma*0.35);
      const d2 = z.reduce((s,vi,j)=>{ const diff=vi - x[j]; return s+diff*diff; },0);
      const w = Math.sqrt(Math.exp(-d2/(kernelWidth*kernelWidth)));
      const y = (typeof predictFn==='function') ? predictFn(z) : dot(z, Array(M).fill(1/M));
      Xli.push([1].concat(z));
      yli.push(y);
      wli.push(w);
    }
    const betaLime = weightedLeastSq(Xli, yli, wli, 1.0);
    const limeIntercept = betaLime[0];
    const limeCoefs = betaLime.slice(1);

    const shap = {}; const lime = {};
    for(let i=0;i<M;i++){ shap[names[i]]=shapArr[i]; lime[names[i]]=limeCoefs[i]; }

    // ---- Narrative ----
    const narrative = makeNarrative(names, shapArr, limeCoefs, x, baseline, pred, fBaseline, opts.domain||'generic');

    return {
      prediction: pred,
      baseline: {value: fBaseline, vector: baseline},
      shap_values: shap,
      shap_array: shapArr,
      lime_values: lime,
      lime_array: limeCoefs,
      intercept_shap: intercept,
      intercept_lime: limeIntercept,
      fidelity: {shap_additive_error: Math.abs(fFull-expFull), shap_relative: fidelity, sum_shap: sumShap, expected: expFull},
      stability: {numShap, numLime, kernelWidth},
      feature_names: names,
      narrative,
      meta: {M, domain:opts.domain||'generic', timestamp: new Date().toISOString(), lcg: '20260813→189831298 idx3820 triple[11205,19448,14209]'}
    };
  }

  function topK(names, vals, k=3){
    const idx = vals.map((v,i)=>[Math.abs(v),i,v]).sort((a,b)=>b[0]-a[0]).slice(0,k);
    return idx.map(([abs,i,v])=>({name:names[i], val:v, abs, idx:i}));
  }

  function makeNarrative(names, shapArr, limeArr, x, baseline, pred, basePred, domain){
    const delta = pred-basePred;
    const tops = topK(names, shapArr, 3);
    const limeTops = topK(names, limeArr, 3);
    const fmt = (v)=> (v>=0?`+${v.toFixed(2)}`:`${v.toFixed(2)}`);
    const shapStr = tops.map(t=>`${t.name} (SHAP ${fmt(t.val)})`).join(', ');
    const limeCheck = limeTops[0] ? `LIME locally ${limeTops[0].name} ${limeTops[0].val>=0?'pushes up':'pulls down'} (${fmt(limeTops[0].val)})` : '';
    // domain-specific POV sentence
    const pov = {
      owner: domain.includes('hoops') ? `Ownership view: this projects ${fmt(delta)} wins above baseline because ${shapStr}.` : domain.includes('gridiron') ? `Cap view: ${fmt(delta)} points above replacement driven by ${shapStr}.` : domain.includes('pitch') ? `Run-value view: ${fmt(delta)} above park-neutral driven by ${shapStr}.` : domain.includes('equities') ? `Edge view: ${fmt(delta)} expected 6M fwd vs random because ${shapStr}.` : `Unified: ${fmt(delta)} lift from ${shapStr}.`,
      player: `Fit view: local shape ${limeCheck} — ${tops.length>1?`secondary ${tops[1].name} ${fmt(tops[1].val)} suggests situational upside`:'stable across neighborhood'}.`,
      brand: `Story view: headline driver is ${tops[0]?.name||'unknown'} (${fmt(tops[0]?.val||0)}) — marketability lifts when ${limeTops[0]?.name||'context'} aligns.`,
      dfs: `DFS view: exploitable if ${tops[0]?.name||'top'} ${tops[0]&&(tops[0].val>0)?'remains mispriced':'is faded'} — SHAP sum ${fmt(shapArr.reduce((a,b)=>a+b,0))} → pred ${pred.toFixed(2)} vs base ${basePred.toFixed(2)}.`
    };
    const generic = `This pick projects ${fmt(delta)} vs baseline (${basePred.toFixed(2)}→${pred.toFixed(2)}) because ${shapStr}. ${limeCheck}, ${pov.player} ${pov.owner}`;
    return {owner:pov.owner, player:pov.player, brand:pov.brand, dfs:pov.dfs, generic, tops, limeTops, delta};
  }

  // expose
  root.Explainer = {explainPrediction, shapKernel, weightedLeastSq, comb};
  if(typeof module!=='undefined' && module.exports) module.exports = root.Explainer;
})(typeof self!=='undefined'?self: (typeof window!=='undefined'?window:globalThis));
